##Mean field Markov Random field with extended spatial parameters EM procedure.
The goal of the algorithm presented in this project is to estimate the parameters of a HMRF model with extended spatial dependencies through an EM procedure using mean field approximation (manuscript submitted for publication)

`Jean-Baptiste Pettit 2014`

`European Bioinformatics Institute`

 NOTICE OF LICENSE - This source is subject to the Academic Free License (AFL 3.0) that is bundled with this package in the file LICENSE_AFL.txt.

## COMPILING THE CODE
The implementation is written in C. a make file is available but has only been tested on a Linux platform. If you are having troubles compiling in other environnement please report an issue on the project Github page (https://github.com/jbogp/MRF_Platynereis_2014/issues)

## BASIC USAGE MANUAL
### REQUIRED FILES
To function, the algorithm requires at least 2 files :
 - The dataset file which contains binarised expression data for each considered point with the following organisation:
   - Each line is a datapoint
   - The first column is a unique identifier for each line
   - The next M columns are the binary gene expression for all the genes considered (0 if not expressed 1 if expressed)
   - The charater to separate columns is a tabulation
   - Note that no missing data is allowed and all points must have expression information for all genes
   
 - The neighbouring graph file :
   - Each line is a datapoint (in the same order as the dataset file)
   - The first colums represents the number of neighbours (u) for this data point
   - The u next columns represent the line number of the neighbours. For example if the point on line 1 has 3 neughbours, the points on line 5, 7 and 12, the file should start with : `3	5	7	12`
   
### INITIALIZATION FILE 
This file is not required as a random intialization can be generated by the algorithm (see the PARAMETERS section for details). However if you wish to initialize the clustering with a specific classification, the format is the follwing :
- Each line is a datapoint (in the same order as the dataset file)
- Each line has only one column that is a number between 1 and K (number of desired clusters) corresponding to the clusters in with the point should initialy be clustered in.

### PARAMETERS INPUT 
The parameters for the algorithm are as follow ([parameter name] indicates the parameter is mandatory {parameter name} indicates it's optional) [param 1 | param2] indicates a choice
- [path to data_file]
- [path neighbouring file]
- ["rand" | path to initialisation file]
- [initial value for beta]
- [number of clusters K]
- [result folder]
- [outputFileName]
- [number of clusters changed from one iteration to the next to assume convergence]
- {"fixed"} (if present, beta will be fixed to initial value instead of being estimated)

Example command : `./EM data/binary_86_genes.tab data/neighbouring_graph.nei rand 0 10 myResultFolder myResultFile 50`

### OUTPUT FILES
The algorithm produces 4 files when convergence is reached
- outputFileName.csv contains the clustering results in the same format as the initialization file
- outputFileName.thetas contains a table with the final values of the theta parameters
- outputFileName.clust contains a table with the final value of beta and the number of points for each cluster
- outputFileName.summary contains information about the run, namely the numbers of clusters, the final log likelihood value and the number of iterations needed to reach convergence
